### PR TITLE
Config files

### PR DESCRIPTION
**Pull request by miharenthas**
Hi, if you're interested, I modified the config files templates slightly to append to the system paths instead of rewriting them --which can cause troubles sometimes...

I also can't really test for csh, so I might have broken that particular file.

### CODE REVIEW

**Comment by MohammadAlTurany** [comment]
Hi @miharenthas 
if you add   _-DUSE_PATH_INFO=TRUE_   when calling cmake, you get exactly this.
So the functionality is there no need for this modification.

**Comment by miharenthas** [comment]
Tried: it still doesn't seem to append to $PATH (and similar), as in export PATH=$PATH:/something/is/here/", but still uses an expansion of the system path (and linkers path).

```
export PATH="/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/gatto/.local/bin:/home/gatto/bin:/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/gatto/.local/bin:/home/gatto/bin:/usr/local/FAIR/FairSoft/build-root6-gcc-7.2.1/bin/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/gatto/.local/bin:/home/gatto/bin:/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/home/gatto/.local/bin:/home/gatto/bin:/usr/local/FAIR/FairSoft/build-root6-gcc-7.2.1/bin:/usr/local/FAIR/FairSoft/build-root6-gcc-7.2.1/bin"
```
Also, it seems to include paths that are not there: I have no binary in my home (and no ~/bin folder!) and it's not this script's business to nose in there without me explicitly telling that there's something of interest in ~.

What I'm trying to achieve is appending the FairThings extensions to the system path as it's found in the current shell --I might have added something in that particular session, and if I then want to use FairRoot or its derivative the paths are rewritten and I have to type again... no big deal, but still if it can be avoided...

**Comment by dennisklein** [comment]
Actually, `-DUSE_PATH_INFO=TRUE` captures the environment at build time, while this patch will prepend the user environment at run time.

File: cmake/scripts/config.csh.in [hunk 1]
```diff
@@ -45,7 +45,7 @@ if ( "$same_system" == "1" ) then
   setenv VMCWORKDIR "@VMCWORKDIR@"
   setenv FAIRLIBDIR "@FAIRLIBDIR@"
   setenv PYTHONPATH "@MY_PYTHONPATH@"
-  setenv DYLD_LIBRARY_PATH "@MY_DYLD_LIBRARY_PATH@"
-  setenv LD_LIBRARY_PATH "@MY_LD_LIBRARY_PATH@"
-  setenv PATH "@MY_PATH@"
+  setenv DYLD_LIBRARY_PATH $DYLD_LIBRARY_PATH:"@MY_DYLD_LIBRARY_PATH@"
+  setenv LD_LIBRARY_PATH $LD_LIBRARY_PATH:"@MY_LD_LIBRARY_PATH@"
+  setenv PATH $PATH:"@MY_PATH@"
```
Comment (dennisklein) [inline cmake/scripts/config.csh.in]: IMHO it should be the other way around. `@MY_PATH@` in front of `$PATH`. Same for the library paths.

Comment (miharenthas) [reply cmake/scripts/config.csh.in]: If we're using libraries and binaries that could be shadowed by the system's ones (and that would create a problem), yes...

**Comment by dennisklein** [comment]
@MohammadAlTurany @fuhlig1 What was the use case for `USE_PATH_INFO`? In my CMake modernization PR I have currently dropped this switch, because I couldn't think of any use cases where I want to burn the build time environment into my installation artefacts. My replacement matches the approach of this PR, just prepends our paths.

**Comment by fuhlig1** [comment]
Hi Lorenzo,

we will not change the setup of the path since it is needed only in very special cases and for these cases we have the option -DUSE_PATH_INFO=TRUE. And yes we want to have the configuration at build time to be also available at runtime. The information about the PATH and the LD_LIBRARY_PATH is only needed at compile time if you have an installation at a rather strange place. For example this is the case if you have to use a non-system compiler which is installed somewhere in the file system. You need this setup at compile time such that your compilation of FairRoot works and you need the same environment at run time because otherwise the compiler dependent libraries (like libstdc++) are not found at run time. So  we implemented the setup of PATH and LD_LIBRARY_PATH to good reasons the way it is implemented. This was done to avoid that people use a complete spoiled PATH and LD_LIBRARY_PATH settings which then transfer all kind of crap (like several different ROOT installations) to the run time environment. Have a lot of fun to find problems when you use the root binary from one installation and libraries from another.
Looking at the export command you ave posted you must have called some setup script several times. I find at least 3 duplicates of all the paths.

If you want to have something special please add the needed paths to your config script after the installation and you are done.

Best regards

Florian

**Comment by dennisklein** [comment]
> If you want to have something special please add the needed paths to your config script after the installation and you are done.

So basically I have to run `source ~/.bash_profile` again after sourcing `config.sh` ... e.g. when using a central FairRoot installation.

**Comment by miharenthas** [comment]
Dear Florian,

The path I posted above was generated by CMake in a fresh shell --and, sure enough, it picked up locations that _don't even exist in my file system, let alone my *PATH's_. That's a deeper problem than "oh, you're rewriting my environment variables and messing my shell up". And one I'm not paid to solve.

But more importantly: Fair{Root,Soft} is not the only bit of software that lives on my computer and, astonishingly, it will not be the last one I'll ever install. Other things will come and some may do their own addition to my environment: if you _rewrite_ the environment, those might break and there's no reason for that. If you {ap,pre}pend, they won't. In particular, if you _pre_pend stuff, as @dennisklein suggested above, you can shadow the system binaries and libraries in favour of whatever you need to use.
Or I would want to add stuff for a particular session to my LD_LIBRARY_PATH and and there's no reason whatsoever why that should be discarded for the sake of making FairRoot reachable, and then I have to dig in my history to retrieve them.

Quite frankly, the environment setup should be done _with an environment module_ whenever possible, and not a shell script, so that FairRoot addition could also be removed without the need of resetting the shell --and you'd not spam your paths if you tried to setup the environment more than once.
Also, with an environment module, you could _require_ another module (for example, that loads a compiler), if you need it --indeed, I noticed that my paths were being screwed up by config.sh when I needed to switch among compilers, something I was doing with environment modules, and by loading a module _and then_ $ . config.sh everything broke.

But hey, the good thing of git is that everyone can have their own branches. I'll keep having mine.

Cheers,
L.

**Comment by dennisklein** [comment]
I have a proposal: How about we just add a second version of the config script, that does it the way it is proposed here and leave the current one as it is?

**Comment by miharenthas** [comment]
It'd be also possible to add a command line option to the script, so `$ . config.sh` rewrites, whereas `$ . config.sh --prepend` would prepend the paths. Or append, with another option.

**Comment by miharenthas** [comment]
There yah go: now the tip of the branch contains the file templates with the option selection:
-a ( and --append for bash ) to append the path
-p ( and --prepend for bash) to prepend the path
without options, the behaviour is business as usual: the paths are rewritten.

**Comment by dennisklein** [comment]
Somehow the refs are screwed, trying with a new PR

**Pull request closed without merging**